ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -0,0 +1,147 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+(**
+ * @winbind group Access Control
+*)
+module D = Debug.Make (struct
+  let name = "extauth_plugin_ADwinbind"
+end)
+
+open D
+
+module Winbind = struct
+  let winbind_service = "winbind"
+
+  let start ~timeout ~wait_until_success =
+    Xapi_systemctl.start ~timeout ~wait_until_success winbind_service
+
+  let stop ~timeout ~wait_until_success =
+    Xapi_systemctl.stop ~timeout ~wait_until_success winbind_service
+
+  let init_service ~__context =
+    debug "init_service To be implemented in CP-36085"
+end
+
+module AuthADWinbind : Auth_signature.AUTH_MODULE = struct
+  (* subject_id Authenticate_username_password(string username, string password)
+
+      Takes a username and password, and tries to authenticate against an already configured
+      auth service (see XenAPI requirements Wiki page for details of how auth service configuration
+      takes place and the appropriate values are stored within the XenServer Metadata).
+      If authentication is successful then a subject_id is returned representing the account
+      corresponding to the supplied credentials (where the subject_id is in a namespace managed by
+      the auth module/service itself -- e.g. maybe a SID or something in the AD case).
+      Raises auth_failure if authentication is not successful
+  *)
+
+  let authenticate_username_password username password =
```

Comment:
I suggest to use labels because you have two strings here: `~username` `~password`

Reply:
I believe Lin is implementing a pre-existing interface here, so it's probably not worth changing this